### PR TITLE
introduce ServiceLogs API

### LINKED ISSUE #425: fix logs
* currently `service logs` command doesn't give provide proper logs. for the service-webhook it only shows first printed message `Example app listening on port 3000!`

* `-d` option of service logs command does not complain for non-existent dependencies.

this can re-produced by [current](https://github.com/mesg-foundation/core/commit/150fcbed789c2162bfe6ed2b4d3515e365261e22) version of dev

referenced by https://github.com/mesg-foundation/core/pull/414#issuecomment-418093285

### PR DESCRIPTION
**Pull request by ilgooz**
closes #425 because I haven't experimented this problem within this PR.

### CODE REVIEW

**Comment by ilgooz** [comment]
good to review

File: cmd/service/logs.go [hunk 1]
```diff
@@ -23,34 +27,156 @@ mesg-core service logs SERVICE_ID --dependency DEPENDENCY_NAME`,
...
 	<-xsignal.WaitForInterrupt()
 }
 
-func showLogs(serviceID string, dependency string) func() {
-	reply, err := cli().GetService(context.Background(), &core.GetServiceRequest{
-		ServiceID: serviceID,
+// dependencyLogs keeps dependency info and corresponding std & err log streams.
+type dependencyLogs struct {
+	Dependency      string
+	Standard, Error *logReader
```
Comment (krhubert) [inline cmd/service/logs.go]: rename stdout stderr

File: service/dependency.go [hunk 2]
```diff
@@ -49,3 +53,21 @@ func (s *Service) DependenciesFromService() []*DependencyFromService {
...
+// error logs.
+func (d *DependencyFromService) Logs() (rstd, rerr io.ReadCloser, err error) {
+	var reader io.ReadCloser
+	reader, err = defaultContainer.ServiceLogs(d.namespace())
+	if err != nil {
+		return nil, nil, err
+	}
+	sr, sw := io.Pipe()
+	er, ew := io.Pipe()
+	go func() {
```
Comment (krhubert) [inline service/dependency.go]: make sure the go routine is finished when connection is closed

File: cmd/service/logs.go [hunk 3]
```diff
@@ -23,34 +27,156 @@ mesg-core service logs SERVICE_ID --dependency DEPENDENCY_NAME`,
...
 	return func() {
-		for _, reader := range readers {
-			reader.Close()
+		for _, c := range append(rstds, rerrs...) {
+			c.Close()
+		}
+	}
+}
+
+// logReader implements io.Reader to combine log data chunks being received
```
Comment (krhubert) [inline cmd/service/logs.go]: Move this out of cmd package

File: api/log.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package api
```
Comment (krhubert) [inline api/log.go]: combine with log-LOGGER.GO

File: interface/grpc/core/logs.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+func newLogPiper(typ LogData_Data_Type, dependency string, stream io.ReadCloser,
+	results chan logChunk) *logPiper {
+	r := &logPiper{
+		Type:       typ,
+		Dependency: dependency,
+		Stream:     stream,
+		Chunks:     results,
+		closing:    make(chan struct{}, 0),
+	}
+	go r.run()
```
Comment (krhubert) [inline interface/grpc/core/logs.go]: move log pieper to log_piper.go

Comment (krhubert) [reply interface/grpc/core/logs.go]: log piper shouldnt start go routines

File: service/dependency.go (hunk 2 above)
Comment (ilgooz) [reply service/dependency.go]: _StdCopy_ returns when there is an error writing to _sw_ or _ew_ which will be caused by calling _Close()_ on _sr_ or _er_.

so yes, goroutine finishes when one of the readers is closed

**Comment by ilgooz** [comment]
good to review

File: commands/commands.go [hunk 6]
```diff
@@ -25,8 +26,7 @@ type ServiceExecutor interface {
 	ServiceDeploy(path string) (id string, valid bool, err error)
 	ServiceListenEvents(id, eventFilter string) (chan *core.EventData, chan error, error)
 	ServiceListenResults(id, taskFilter, outputFilter string, tagFilters []string) (chan *core.ResultData, chan error, error)
-	ServiceLogs(id string) (io.ReadCloser, error)
-	ServiceDependencyLogs(id string, dependency string) ([]io.ReadCloser, error)
+	ServiceLogs(id string, dependencies ...string) (logs []*provider.Log, closer func(), err error)
```
Comment (krhubert) [inline commands/commands.go]: update commands_mock_test with new interface

File: commands/service_logs.go [hunk 7]
```diff
@@ -30,25 +34,74 @@ mesg-core service logs SERVICE --dependency DEPENDENCY_NAME`,
...
 	return nil
 }
+
+func showLogs(e ServiceExecutor, serviceID string, dependencies ...string) (closer func(), err error) {
+	logs, closer, err := e.ServiceLogs(serviceID, dependencies...)
+	if err != nil {
+		return nil, err
+	}
+
+	dependencies = []string{}
```
Comment (krhubert) [inline commands/service_logs.go]: if len(dependencie)== 0 {....

File: commands/service_logs.go [hunk 8]
```diff
@@ -30,25 +34,74 @@ mesg-core service logs SERVICE --dependency DEPENDENCY_NAME`,
...
+
+	return closer, nil
+}
+
+// dependencyPrefixes returns dependency key, log prefix pair.
+func dependencyPrefixes(dependencies []string) (prefixes map[string]string) {
+	// maxCharLen is the char length of longest dependency key.
+	var maxCharLen int
+
+	prefixes = make(map[string]string, 0)
```
Comment (krhubert) [inline commands/service_logs.go]: remove 0 or change to len(dependencies)

File: commands/provider/service_provider.go [hunk 9]
```diff
@@ -118,23 +119,75 @@ func (p *ServiceProvider) ServiceListenResults(id, taskFilter, outputFilter stri
 	return resultC, errC, nil
 }
 
+type Log struct {
+	Dependency      string
+	Standard, Error *chunker.Stream
```
Comment (krhubert) [inline commands/provider/service_provider.go]: Change to stdout and stderr

File: commands/provider/service_provider.go [hunk 10]
```diff
@@ -58,7 +57,9 @@ func (p *ServiceProvider) ServiceDeleteAll() error {
 func (p *ServiceProvider) ServiceDelete(ids ...string) error {
 	var errs xerrors.Errors
 	for _, id := range ids {
+		fmt.Println(11, id)
```
[the remark this answers is not in the document]
Comment (ilgooz) [reply commands/provider/service_provider.go]: already done

**Pull request merged**